Challenge Page

lucapircher opened the pull request:
# Description

- adds all the supporting infrastructure for using the app router and the pages router
   - graphql authentication
   - authentication
   - application configuration
   - adjusts the navigation, breadcrumbs, etc. so that they are compatible with the pages and the app raut
- creates a dedicated page for the challenges using the nextjs app router

Closes https://github.com/a11yphant/issues/issues/46

Review App: https://a11yphant-158-merge.cluster.a11yphant.com

## Definition of Done

### General

- [ ] Code Review
- [x] Test Coverage is equal or higher
- [x] Update ticket on the TODO board
- [x] (if .env file changes) Notify other team members

### Site

- [x] Works in Firefox, Chrome and Safari
- [x] No W3C Errors (Wave Check)
- [x] Focus, Hover & Active Styles are present
- [x] Correct Tab Order
- [x] All relevant elements are focusable
- [x] Screen reader only reads relevant infos (no SVGs, ...)

lucapircher committed:
feat(site): create app router challenge page

lucapircher committed:
feat(site): reset api package lockfile

lucapircher committed:
feat(site): reset tsconfig formatting

lucapircher committed:
chore(site): update env variables for tests

lucapircher committed:
feat(site): update plausible config

lucapircher committed:
feat(site): add encoding package

lucapircher committed:
feat(site): add redirect for old challenge urls

lucapircher committed:
feat(site): add redirect for old challenge urls

lucapircher committed:
feat(site): update test env variables and use type export

lucapircher committed:
feat(site): fix error page

lucapircher committed:
feat(site): add sharp

lucapircher committed:
feat(site): remove runtime config

lucapircher committed:
feat(site): warn for missing configuration instead of throwing

lucapircher committed:
chore(site): remove env variables for linting

lucapircher committed:
chore(site): add env variables for tests

lucapircher committed:
feat(site): update login and logout functionality

lucapircher committed:
test(e2e): increase login timeout

lucapircher committed:
feat(site): fix typo in flash message

lucapircher committed:
feat(site): fix breadcrumb rendering

lucapircher committed:
feat(site): use next script tag

lucapircher committed:
test(site): add tets

lucapircher committed:
refactor(site): move component to app directory

lucapircher committed:
test(site): add test for root layout

lucapircher commented on services/site/src/app/challenge/[challengeSlug]/page.tsx:
```diff
@@ -0,0 +1,65 @@
+    notFound();
+  }
+
+  const firstUnfinishedLevel =
```
Other locations were removed.

lucapircher commented on services/site/src/middleware.ts:
```diff
@@ -1,10 +1,48 @@
 import { NextRequest, NextResponse } from "next/server";
 
-export default function middleware(req: NextRequest): NextResponse {
+export default async function middleware(req: NextRequest): Promise<NextResponse | void> {
```
Refactored

lucapircher commented on services/site/src/middleware.ts:
```diff
@@ -1,10 +1,54 @@
+import { getConfig } from "./lib/config/rsc";
 
-export default function middleware(req: NextRequest): NextResponse {
+export default async function middleware(req: NextRequest): Promise<NextResponse | void> {
```
Refactored

lucapircher commented on services/site/src/middleware.ts:
```diff
@@ -1,10 +1,54 @@
+import { getConfig } from "./lib/config/rsc";
 
-export default function middleware(req: NextRequest): NextResponse {
+export default async function middleware(req: NextRequest): Promise<NextResponse | void> {
```
Refactored

lucapircher committed:
refactor(site): refactor route matching for breadcrumbs

lucapircher commented on services/site/src/components/breadcrumbs/routes.tsx:
```diff
@@ -1,66 +1,72 @@
-  "/challenge/[challengeSlug]": {
-    getBreadcrumbInfo: async (urlParams: { challengeSlug: string }, apolloClient) => {
-      const { challengeSlug } = urlParams;
+export function match(key: string): RouteInfo | null {
```
Refactored

lucapircher commented on services/site/src/components/breadcrumbs/routes.tsx:
```diff
@@ -1,66 +1,72 @@
+    };
+  }
+
+  return null;
```
Refactored

lucapircher committed:
feat(site): fix heading structure and checkmark

lucapircher committed:
feat(site): nprogress for app router

lucapircher committed:
feat(site): add base metadata for pages

lucapircher committed:
refactor(site): inline component for level list item

lucapircher commented on services/site/src/app/challenge/[challengeSlug]/page.tsx:
```diff
@@ -0,0 +1,74 @@
+    <>
+      <Navigation displayBreadcrumbs />
+      <main className="h-full box-border max-w-screen-3xl mx-auto">
+        <div className="mx-8 py-8 h-main max-w-screen-3xl sm:mx-12 lg:mt-12 lg:mx-24">
+          <h1 className="mb-8 h2 md:h1">{challenge?.name}</h1>
+          <p className="prose" dangerouslySetInnerHTML={{ __html: sanitizeHtml(challenge.introduction) }} />
+
+          <ul className="gap-4 pt-8 grid grid-cols-1 gap-y-4 sm:grid-cols-2 lg:grid-cols-3 2xl:grid-cols-4">
+            {challenge.levels.map((level) => {
+              const isFirstUnfinishedLevel = level.id === firstUnfinishedLevel?.id;
+              return (
+                <li className="m-0 p-0" key={level.id}>
+                  <Link
+                    href={`/challenge/${challengeSlug}/level/${String(level.order).padStart(2, "0")}`}
+                    className={clsx(
+                      "relative block border border-solid rounded-lg px-4 py-3 w-full h-18",
+                      "hover:bg-primary-dark hover:border-primary-dark",
+                      !isFirstUnfinishedLevel && "border-grey-dark",
+                      isFirstUnfinishedLevel && "border-primary bg-primary",
+                    )}
+                  >
+                    <span className="h3 text-base text-grey block">Level {String(level.order).padStart(2, "0")}</span>
+                    <span className="font-normal text-grey-middle mb-0 block">{level.__typename === "QuizLevel" ? "Quiz" : "Coding"}</span>
+                    {level.status === LevelStatus.Finished && <Check className="h-7 w-10 absolute top-4 right-5 text-light" />}
+                  </Link>
+                </li>
+              );
+            })}
+          </ul>
```
In the modal this was structured as multiple level 3 headings each containing one paragraph with the word "Coding" or "Quiz" the only content within that heading. "A list of levels" makes conceptional more sense to me.

dnikub replied:
I understand where you are coming from but am unsure here. It could be easier to jump to a specific level via the heading with a SR when you list all headings with the SR shortcut. But I am unsure if it is really needed here, maybe adding a skip link to unsolved levels is the better option.

lucapircher replied:
Maybe to add a bit more context: I checked the document outline and I noticed that we have the headings and then only a single word ("Coding" or "Quiz") as the content. This made me thinking about whether this is the best way to structure this. I understand that it's mainly for jumping to the levels and not really for the document structure, but it still seemed a bit odd to me.

Maybe the following would provide the best experience: Keep the levels as a list, but add a visually hidden heading so that it's possible to jump to the levels. Additionally we could then add a quick link to continue with the next unfinished level. I think that would be a great addition.

Currently you'd have to go through all levels anyways when using a SR, since we don't indicate the status in the heading. For this reason I think we should tackle this outside of this MR.

lucapircher commented on services/site/src/app/layout.tsx:
```diff
@@ -0,0 +1,60 @@
+
+export default RootLayout;
+
+export const metadata: Metadata = {
```
This metadata will be the default value for each of the pages in the app directory. So we only have to specify the values that should be different in the individual pages. That will allow us to get rid of the duplication that we currently have in the pages.

dnikub replied:
Can you please use this new way of displaying metadata on all pages we have where meta info for SEO is used. If that was done already, nice! its just so many changes maybe I overlooked it.

lucapircher replied:
Unfortunately we can't use this for the pages we already have, since this feature is only available with the "app router" and those use the "pages router". While migrating the existing pages to app router pages we can also migrate the metadata and remove the duplication there.

lucapircher commented on services/site/src/components/Navigation.tsx:
```diff
@@ -19,10 +22,16 @@ export interface NavigationProps {
+
+  const logout = async (): Promise<void> => {
+    await triggerLogout();
+    router.push("/");
```
We now have to explicitly refresh the page, since the statically rendered content of the page for a logged in user doesn't match the content that should be shown for a anonymous user.

lucapircher commented on services/site/src/components/breadcrumbs/Breadcrumbs.tsx:
```diff
@@ -27,7 +27,7 @@ const Breadcrumbs: React.FunctionComponent = () => {
       isMounted = false;
     };
-  }, [apolloClient, router, router.pathname]);
+  }, [apolloClient, pathname, JSON.stringify(params)]);
```
`useParams` returns a new object each time it's called. Using that as a trigger for an effect leads to a rendering infinity loop. As a fix we JSON encode the object so that the value changes only if one of the values changes.

lucapircher commented on services/site/src/lib/config/index.tsx:
App directory doesn't support the server and client configuration that was available in Next 12 (and Next 13 pages) anymore. For that reason we had to rebuild a similar thing for the client side configuration.

lucapircher commented on services/site/src/middleware.ts:
```diff
@@ -1,10 +1,73 @@
+export default async function middleware(req: NextRequest): Promise<NextResponse | null> {
+  const middlewares = [forwardGraphqlRequests, redirectChallengeOverlayUrls, authentication];
+
+  for (const middleware of middlewares) {
```
NextJS supports only a single middleware so we have to create our own middleware concept within that and check which functionality we actually want to execute.
We either forward the request to the gql server, redirect old challenge modal urls, or handle the authentication cookies

lucapircher committed:
feat(site): update page title for challenge page

dnikub commented:
I would add 2 things for improved usability:

1. When a level within the challenge is completed and gets the checkmark icon, I'd add SR text that says this level is already completed. currently, it only says that the link was visited, but that means nothing in this context. 
2. The one level that is coloured purple: I'd add the text "up next"/"start here" when it is the first one, or smth similar. otherwise, it is not completely clear why this one level is colored. Especially when no level was completed yet, is it not initially clear that one can click on the level to start (and we removed the start button).

Otherwise, it looks very good. We should, however, add a new ticket to add a skip link to the newest not completed level so you don't have to tab so much with your keyboard.

dnikub commented on services/site/src/app/challenge/[challengeSlug]/page.tsx:
```diff
@@ -0,0 +1,85 @@
+      <Navigation displayBreadcrumbs />
+      <main className="h-full box-border max-w-screen-3xl mx-auto">
+        <div className="mx-8 py-8 h-main max-w-screen-3xl sm:mx-12 lg:mt-12 lg:mx-24">
+          <h1 className="mb-8 h2 md:h1">{challenge?.name}</h1>
```
I would add a SR text like "Challenge:" to indicate would are on the challenge detail view. otherwise the heading just says "heading" for the headings challenge.

dnikub commented on services/site/src/app/challenge/[challengeSlug]/page.tsx:
```diff
@@ -0,0 +1,85 @@
+      <Navigation displayBreadcrumbs />
+      <main className="h-full box-border max-w-screen-3xl mx-auto">
+        <div className="mx-8 py-8 h-main max-w-screen-3xl sm:mx-12 lg:mt-12 lg:mx-24">
+          <h1 className="mb-8 h2 md:h1">{challenge?.name}</h1>
```
Also, I am missing the indicator of the challenge difficulty here. It was present in the modal.

lucapircher replied:
Oh yeah that went missing somewhere along the way. I'll add it back.

lucapircher replied:
@dnikub I added it back, but I am not really sure about the size and spacing for it. Please take a look.

dnikub replied:
You are missing SR description here. Please add a sr-only p tag with the text "Difficulty easy" inside the div that has the thee spans. Then it is in alignment with Other sections on our website.

lucapircher commented:
Thanks for reviewing.

> I would add 2 things for improved usability:
> 
> 1. When a level within the challenge is completed and gets the checkmark icon, I'd add SR text that says this level is already completed. currently, it only says that the link was visited, but that means nothing in this context.

Ah yeah that makes a lot of sense, I'll add that.

> 2. The one level that is coloured purple: I'd add the text "up next"/"start here" when it is the first one, or smth similar. otherwise, it is not completely clear why this one level is colored. Especially when no level was completed yet, is it not initially clear that one can click on the level to start (and we removed the start button).

I'll also add that. Should we add that to the main text in the card? I added it in the devtools in screenshot below and I kinda like it.

<img width="561" alt="image" src="https://github.com/a11yphant/a11yphant/assets/6804146/6b11c278-9851-457d-805b-276e1bdb145b">

> Otherwise, it looks very good. We should, however, add a new ticket to add a skip link to the newest not completed level so you don't have to tab so much with your keyboard.

Ah sorry, I started replying to the other comment before I saw this. Yeah doing that in a dedicated ticket is fine by me and it would be a great addition.

lucapircher commented:
@thomasdax98 It would be great if you could also do a review for this one, since it will act as the foundation for future pages.

lucapircher committed:
feat(site): add indicator for where to continue with the challenge

lucapircher committed:
feat(site): add difficulty indicator

lucapircher committed:
Merge branch 'develop' into feature/challenge-page

lucapircher committed:
chore(deps): add jest types

lucapircher committed:
Merge branch 'develop' into feature/challenge-page

lucapircher commented on services/site/src/app/layout.tsx:
```diff
@@ -0,0 +1,60 @@
+  const plausibleUrl = process.env.SITE_PLAUSIBLE_BASE_URL;
+  const domain = process.env.SITE_HOST;
+
+  loadDevMessages();
+  loadErrorMessages();
```
Check if this needs to be disabled for production builds

lucapircher committed:
feat(site): update apollo rsc pacakge

lucapircher committed:
feat(site): add babel transform for superjson

lucapircher committed:
feat(site): update experimental apollo client package

lucapircher committed:
feat(site): conditionaly load dev messages

dnikub commented on services/site/src/app/challenge/[challengeSlug]/page.tsx:
```diff
@@ -0,0 +1,110 @@
+                    <span className="font-normal text-grey-middle mb-0 block">{level.__typename === "QuizLevel" ? "Quiz" : "Coding"}</span>
+                    {level.status === LevelStatus.Finished && (
+                      <>
+                        <span className="sr-only">You have already completed this level.</span>
```
I would shorten the text here as it gets very tiring with the long sentence when you already completed 6 Levels and have to go through every one.

```suggestion
                        <span className="sr-only">Completed.</span>
```

dnikub commented on services/site/src/app/challenge/[challengeSlug]/page.tsx:
```diff
@@ -0,0 +1,110 @@
+            <span className="sr-only">Challenge: </span>
+            {challenge?.name}
+            <div className={clsx("inline-block")}>
+              <DifficultyIcon className={clsx("w-5 h-12 inline-block")} firstClassName="ml-4" />
```
We miss an accessible name here. Within the component we should have a sr-only p that says "Difficulty Easy" for example. Then it is like it is in the Challenge car that gets displayed on the homepage. Maybe we need to add it to the component and remove it from the challenge card. Depends on how the code currently looks.

dnikub commented on services/site/src/app/challenge/[challengeSlug]/page.tsx:
```diff
@@ -0,0 +1,110 @@
+                    <span className="h3 text-base text-grey block">
+                      {getTitlePrefix()}Level {String(level.order).padStart(2, "0")}
+                    </span>
+                    <span className="font-normal text-grey-middle mb-0 block">{level.__typename === "QuizLevel" ? "Quiz" : "Coding"}</span>
```
When the Level has the purple background color because it is the next to complete level, the Quiz/Coding Text needs to be white, otherwise we have a contrast error.

lucapircher committed:
feat(site): improve sr texts

lucapircher committed:
test(site): update tests for difficulties
